pkg/transparentproxy/iptables/setup.go

```diff
@@ -17,11 +18,110 @@ func Setup(ctx context.Context, cfg config.InitializedConfig) (string, error) {
+
+	if err := Cleanup(ctx, cfg); err != nil {
+		return "", errors.Wrap(err, "cleanup failed during setup")
+	}
+
 	return builder.RestoreIPTables(ctx, cfg)
 }
 
-func Cleanup(_ config.InitializedConfig) (string, error) {
-	return "", errors.New("cleanup is not supported")
+// Cleanup removes iptables rules and chains related to the transparent proxy
+// for both IPv4 and IPv6 configurations. It calls the internal cleanupIPvX
+// function for each IP version, ensuring that only the relevant rules andran
+// chains are removed based on the presence of iptables comments. If either
+// cleanup process fails, an error is returned.
+func Cleanup(ctx context.Context, cfg config.InitializedConfig) error {
+	if err := cleanupIPvX(ctx, cfg.IPv4); err != nil {
+		return errors.Wrap(err, "failed to cleanup IPv4 rules")
+	}
+
+	if err := cleanupIPvX(ctx, cfg.IPv6); err != nil {
+		return errors.Wrap(err, "failed to cleanup IPv6 rules")
+	}
+
+	return nil
+}
+
+// cleanupIPvX removes iptables rules and chains related to the transparent
+// proxy, ensuring that only the relevant rules and chains are removed based on
+// the presence of iptables comments and chain name prefixes. It verifies the
+// new rules after cleanup and restores them if they are valid.
+func cleanupIPvX(ctx context.Context, cfg config.InitializedConfigIPvX) error {
+	if !cfg.Enabled() {
+		return nil
+	}
+
+	cfg.Logger.Infof(
+		"starting cleanup of existing transparent proxy rules. Any rule found in chains with names starting with %q or containing comments starting with %q will be deleted",
+		cfg.Redirect.NamePrefix,
+		cfg.Comment.Prefix,
+	)
+
+	// Execute iptables-save to retrieve current rules.
+	stdout, _, err := cfg.Executables.IptablesSave.Exec(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to execute iptables-save command")
+	}
+
+	output := stdout.String()
+	containsTProxyRules := strings.Contains(output, cfg.Redirect.NamePrefix)
+	containsTProxyComments := strings.Contains(output, cfg.Comment.Prefix)
+
+	switch {
+	case !containsTProxyRules && !containsTProxyComments:
+		// If there are no transparent proxy rules or chains, there is
+		// nothing to do.
+		cfg.Logger.Info("no transparent proxy rules detected. No cleanup necessary")
+		return nil
+	case containsTProxyRules && !containsTProxyComments:
+		return errors.New("transparent proxy rules detected, but expected comments are missing. Cleanup cannot proceed safely without comments to identify rules. Please remove the transparent proxy iptables rules manually")
+	}
+
+	// Split the output into lines and remove lines related to transparent
+	// proxy rules and chains.
+	lines := strings.Split(output, "\n")
+	// Remove lines related to transparent proxy rules and chains.
+	// This includes:
+	// 1. Lines starting with "#" (comments).
+	// 2. Lines containing the comment prefix specified in the config.
+	// 3. Lines containing the redirect name prefix specified in the config.
```

Comment:
The code below is really good enough 👍 I don't think this comment brings any new context

Reply:
removed